src/Cabal2JSON.hs

```diff
@@ -353,24 +361,89 @@ instance HasCodec Mixin where
         <*> requiredField' "include-renaming" .= mixinIncludeRenaming
 
 instance HasCodec LegacyExeDependency where
-  -- TODO: ?
-  codec = undefined
+  codec =
+    object "LegacyExeDependency" $
+      LegacyExeDependency
+        <$> requiredField' "package-name" .= (\(LegacyExeDependency s _) -> s)
+        <*> requiredField' "include-renaming" .= (\(LegacyExeDependency _ vr) -> vr)
 
 instance HasCodec ExeDependency where
-  -- TODO: ?
-  codec = undefined
+  codec =
+    object "ExeDependency" $
+      ExeDependency
+        <$> requiredField' "package-name" .= (\(ExeDependency a _ _) -> a)
+        <*> requiredField' "executable-component-name" .= (\(ExeDependency _ b _) -> b)
+        <*> requiredField' "version-range" .= (\(ExeDependency _ _ c) -> c)
 
 instance HasCodec PkgconfigDependency where
-  -- TODO: ?
-  codec = undefined
+  codec =
+    object "PkgconfigDependency" $
+      PkgconfigDependency
+        <$> requiredField' "pkgconfigName" .= (\(PkgconfigDependency name _) -> name)
+        <*> requiredField' "pkgconfigVersionRange" .= (\(PkgconfigDependency _ version) -> version)
+
+instance HasCodec PkgconfigName where
+  codec = dimapCodec mkPkgconfigName unPkgconfigName codec
+
+instance HasCodec PkgconfigVersionRange where
+  codec = object "PkgconfigVersionRange" $
+    dimapCodec f g $
+      eitherCodec (requiredField' "pc-this-version") $
+        eitherCodec (requiredField' "pc-later-version")$
+          eitherCodec (requiredField' "pc-earlier-version")$
+            eitherCodec (requiredField' "pc-or-later-version")$
+              eitherCodec (requiredField' "pc-or-earlier-version")$
+                eitherCodec ((,) <$> requiredField' "range1" .= fst <*> requiredField' "range2" .= snd)$
+                  eitherCodec((,) <$> requiredField' "range1" .= fst <*> requiredField' "range2" .= snd)
+                    (pure PcAnyVersion)
+    where
+      f = \case
+        Left e -> PcThisVersion e
+        Right (Left e) -> PcLaterVersion e
+        Right (Right (Left e)) -> PcEarlierVersion  e
+        Right (Right (Right (Left e))) -> PcOrLaterVersion  e
+        Right (Right (Right (Right (Left e)))) -> PcOrEarlierVersion e
+        Right (Right (Right (Right (Right (Left(r1,r2)))))) -> PcUnionVersionRanges r1 r2
+        Right (Right (Right (Right (Right (Right (Left (r1, r2))))))) -> PcIntersectVersionRanges r1 r2
+        _ -> PcAnyVersion
+      g = \case
+        PcThisVersion e -> Left e
+        PcLaterVersion e -> Right $ Left e
+        PcEarlierVersion e -> Right $ Right $ Left e
+        PcOrLaterVersion e -> Right $ Right $ Right $ Left e
+        PcOrEarlierVersion e -> Right $ Right $ Right $ Right $ Left e
+        PcUnionVersionRanges r1 r2 -> Right $ Right $ Right $ Right $ Right $ Left(r1, r2)
+        PcIntersectVersionRanges r1 r2 -> Right $ Right $ Right $ Right $ Right $ Right $ Left(r1, r2)
+        PcAnyVersion -> Right $ Right $ Right $ Right $ Right $ Right $ Right ()
```

Comment:
I think i do not really understand how autodocodec works yet, if i have to write such big expressions. I guess I could group these constructors into 3 different categories based on their input parameters. But would this actually help with readability?

Also the corresponding unit test for this instance fails because of blocking mvar. Why?

Reply:
Oh gosh this is terrifying!
Is there no string-representation of this?
LIke `base <= 5 && >= 4.9` 

Blocking on mvar usually means an infinite loop somewhere.

Reply:
I am now using the string representation and parsing from cabal which results in way less code.